addons/hr_timesheet/hr_timesheet.py

```diff
@@ -103,129 +77,118 @@ def on_change_unit_amount(self, cr, uid, id, prod_id, unit_amount, company_id, u
-            raise openerp.exceptions.RedirectWarning(msg, action_id, _('Go to the configuration panel'))
-        emp = emp_obj.browse(cr, uid, emp_id[0], context=context)
-        if emp.journal_id:
-            return emp.journal_id.id
-        else :
-            raise UserError(_('No analytic journal defined for \'%s\'.\nYou should assign an analytic journal on the employee form.') % (emp.name,))
-
-
-    _defaults = {
-        'product_uom_id': _getEmployeeUnit,
-        'product_id': _getEmployeeProduct,
-        'general_account_id': _getGeneralAccount,
-        'journal_id': _getAnalyticJournal,
-        'date': lambda self, cr, uid, ctx: ctx.get('date', fields.date.context_today(self,cr,uid,context=ctx)),
-        'user_id': lambda obj, cr, uid, ctx: ctx.get('user_id') or uid,
-    }
-    def on_change_account_id(self, cr, uid, ids, account_id, context=None):
-        return {'value':{}}
-
-    def on_change_date(self, cr, uid, ids, date):
-        if ids:
-            new_date = self.read(cr, uid, ids[0], ['date'])['date']
-            if date != new_date:
-                warning = {'title':_('User Alert!'),'message':_('Changing the date will let this entry appear in the timesheet of the new date.')}
-                return {'value':{},'warning':warning}
-        return {'value':{}}
-
-    def create(self, cr, uid, vals, context=None):
-        if context is None:
-            context = {}
-        emp_obj = self.pool.get('hr.employee')
-        emp_id = emp_obj.search(cr, uid, [('user_id', '=', context.get('user_id') or uid)], context=context)
-        ename = ''
-        if emp_id:
-            ename = emp_obj.browse(cr, uid, emp_id[0], context=context).name
-        if not vals.get('journal_id'):
-           raise UserError(_('No \'Analytic Journal\' is defined for employee %s \nDefine an employee for the selected user and assign an \'Analytic Journal\'!') % (ename,))
-        if not vals.get('account_id'):
-           raise UserError(_('No analytic account is defined on the project.\nPlease set one or we cannot automatically fill the timesheet.'))
-        return super(hr_analytic_timesheet, self).create(cr, uid, vals, context=context)
-
-    def on_change_user_id(self, cr, uid, ids, user_id):
-        if not user_id:
-            return {}
-        context = {'user_id': user_id}
-        return {'value': {
-            'product_id': self. _getEmployeeProduct(cr, uid, context),
-            'product_uom_id': self._getEmployeeUnit(cr, uid, context),
-            'general_account_id': self._getGeneralAccount(cr, uid, context),
-            'journal_id': self._getAnalyticJournal(cr, uid, context),
-        }}
-
-class account_analytic_account(osv.osv):
+            raise exceptions.ValidationError("This user or employee is not associated to a valid Product")
+    @api.model
+    def _get_employee_unit(self, user_id=None):
+            emp = self.env['hr.employee'].search([('user_id', '=', self.user_id.id or user_id or self.env.uid)])
```

Comment:
How can you have a self.user_id.id on a model method ? You will probably not receive a recordset.

Same for othe default methods.